Mathlib/Computability/DFA.lean

```diff
@@ -163,7 +174,50 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [Set.mem_singleton_iff] at ha' hc'
   substs ha' hc'
   have h := M.evalFrom_of_pow hb hb'
-  rwa [mem_accepts, evalFrom_of_append, evalFrom_of_append, h, hc]
+  rwa [mem_accepts', evalFrom_of_append, evalFrom_of_append, h, hc]
 #align DFA.pumping_lemma DFA.pumping_lemma
 
+section Maps
+
+universe u' v'
+variable {α' : Type u'} {σ' : Type v'} (f : α ≃ α') (g : σ ≃ σ') (s : σ') (a : α') (x : List α')
+
+/-- Transforms the alphabet and states of a DFA. -/
+@[simps]
+def map (f : α' → α) (g : σ → σ') (h : σ' → σ) : DFA α' σ' where
+  step s a := g (M.step (h s) (f a))
+  start := g M.start
+  accept := g '' M.accept
+
+/-- Lifts equivalences on the alphabet and states to an equivalence on DFAs. -/
+def equiv (f : α ≃ α') (g : σ ≃ σ') : DFA α σ ≃ DFA α' σ' where
+  toFun M := M.map f.symm g g.symm
+  invFun M := M.map f g.symm g
+  left_inv M := by simp [map]
+  right_inv M := by simp [map]
+
+@[simp]
+theorem equiv_step : (equiv f g M).step s a = g (M.step (g.symm s) (f.symm a)) := by simp [equiv]
+
+@[simp] theorem equiv_start : (equiv f g M).start = g M.start := by simp [equiv]
+
+@[simp] theorem equiv_accept : (equiv f g M).accept = g '' M.accept := by simp [equiv]
+
+@[simp]
+theorem equiv_evalFrom : (equiv f g M).evalFrom s x = g (M.evalFrom (g.symm s) (x.map f.symm)) := by
+  induction x using List.list_reverse_induction with
+  | base => simp
+  | ind x a ih => simp [ih]
+
+@[simp]
+theorem equiv_eval : (equiv f g M).eval x = g (M.eval (x.map f.symm)) := by simp [eval]
+
+@[simp]
+theorem equiv_accepts : (equiv f g M).accepts = List.map f '' M.accepts := by
```

Comment:
```suggestion
theorem equiv_accepts : (equiv f g M).accepts = List.map f.symm ⁻¹' M.accepts := by
```
Using the preimage of the inverse instead of the image has (usually) better definitional properties.